### PR TITLE
Minigun barrel rotation speed animation config

### PR DESCRIPTION
**Pull request by GoldSloth**
## Description of changes
Added animation modifier to change visual rotation speed of minigun barrels.

## Intended usage in Content Packs/Users of the mod
Config in gun model java (minigunBarrelRotationSpeed).

## Compatibility
Will not affect existing behaviour, including time until the minigun starts shooting.

Default speed =1 which is same rotation speed as present.

## Other
This animation setting needs to be added to the GunType model setting overrides too.

Note I haven't actually tested this, it's just an example for xdeadstokex to review.

### CODE REVIEW

**Comment by xdeadstokex** [comment]
@GoldSloth 
ModelGun.java: line 45-46

public float AngleUpdateSpeed = 2F;
public float spinX=1F, spinY=0F, spinZ=0F;

GunAnimation.java: line 53 and 249

public float AngleUpdateSpeed = 2F;
...
minigunBarrelRotationSpeed += AngleUpdateSpeed;

change from minigunBarrelRotationSpeed +=2F;

RenderGun.java: line 673-675

if(model.AngleUpdateSpeed!=2F){animations.AngleUpdateSpeed=model.AngleUpdateSpeed;
model.AngleUpdateSpeed=2F;}
GL11.glRotatef(animations.minigunBarrelRotation, model.spinX, model.spinY, model.spinZ);



So basically, this is a compromise way to support for custom spin speed and dir:
The first ingame call of the "renderminigun" still use old 2F bcs the "doshoot" func is called first, but after that it should use custom value in your Model (tested sucessfully), also the only downside is that if(), but it only do calculation in the first call only, after that it just do check condition which wont hurt performance much.

**Comment by GoldSloth** [comment]
I was inclined to go with your solution and did start trying to implement this, however it breaks some of the rules of the way the rendering works:
- Model contains the parts e.t.c. and static (don't change at runtime) parameters for how animations should work
- Gun Animations contain the key animation parameters, and these are largely independent of the model animation parameters.
-  RenderGun uses the above two to actually perform the animations

If in RenderGun you were to override the value of rotation speed in GunAnimations, the animation variable is now dependent on model parameters. In GunAnimations.java, there aren't any references to the model - there isn't even access to it. If you had two miniguns with very different rotation speeds, switching between the two would lead to some unexpected results in this case.

I think the solution I've suggested keeps both functions (inc. the rotation point) and is it a bit simpler - what do you think?

(Note I still haven't tested this, been quite busy this week)

**Comment by xdeadstokex** [comment]
@GoldSloth well, my short memory term got the better of me, I did think about gunAnimation being a static one rather than a per model one, but forgot to test 2 different miniguns to confirm it, tested today, well it is a static one, my code still work, just change the if().... to this:

if(animations.AngleUpdateSpeed!=model.AngleUpdateSpeed) animations.AngleUpdateSpeed=model.AngleUpdateSpeed;

which work in different models, obviously the con is always checking condition and when you switch between miniguns, you will see the minor visual change (ex: the gun 1 barrel is spinning fast, switch to gun 2 instantly and you still see the barrel spinning fast despite having a slower spin speed, it is notable when the speed is high but it is not really my code fault, it is an oversight from the Flan code despite my code change, should have reset spin speed everytime you switch gun, it dont even reset with a new gun too lol)

well another method not touching gunAnimation at all is well, a way to reset the value when use custom spin speed but that way is tedious.
Also another way is implement a new free custom animation system but that is for another day :D 

Welp, the indirect update of your is unpleasant to use but it works (also Vector3f is just another layer of codes, but it will not impact performance much) so you can decide what to use in the meantime, the render system of flan mod is flawed at this point, def need improvement.

**Comment by GoldSloth** [comment]
Thanks for your comments on that, I do much prefer the method I suggested because it keeps separation of concerns between the animations, model and renderer - and it's much simpler. Would you be able to test it out? I'll test it as well before merging, but I'd appreciate your input.

I wouldn't worry about using Vectors over normal variables, yeah it'll make barely any difference - the real issues with Flans rendering are code cleanliness or more architectural as ar as I'm aware.

**Comment by xdeadstokex** [comment]
@GoldSloth fix typo:
ModelGun.java:
	public Vector3f minigunBarrelSpinDirection = new Vector3f(1F, 0F, 0F);
	public float minigunBarrelRotationSpeed = 1F;

RenderGun.java:
	GL11.glRotatef(animations.minigunBarrelRotation * model.minigunBarrelRotationSpeed, model.minigunBarrelSpinDirection.x, model.minigunBarrelSpinDirection.y, model.minigunBarrelSpinDirection.z);

otherwise it work, build and run successfully as intended, just typo error.

**Comment by GoldSloth** [comment]
Good spot, thanks :)

**Comment by GoldSloth** [comment]
Can confirm it works, merging :)
https://streamable.com/pkcs40
